code/datums/components/barber.dm

```diff
@@ -453,6 +441,149 @@ TYPEINFO(/datum/component/barber)
+			return TRUE
+
+		if("do_hair")
+			if (isnull(params["style_id"])) // It means we are making a wig
+				if (istype(src, /datum/component/barber/shave))
+					actions.start_and_wait(new/datum/action/bar/barber/shave(src.barbee, src.barber, get_barbery_conditions(src.barbee, src.barber), null, ALL_HAIR), src.barber)
+				else
+					actions.start_and_wait(new/datum/action/bar/barber/haircut(src.barbee, src.barber, get_barbery_conditions(src.barbee, src.barber), null, ALL_HAIR), src.barber)
+
+				if (!barber)
+					return // If there's no barber, it's safe to say we've been disposed of
+
+				src.new_AH.CopyOther(src.barbee.bioHolder.mobAppearance)
+				src.preview.preview_thing.appearance = src.barbee.appearance
+				src.preview.update_appearance(src.new_AH)
+				return TRUE
+
+			var/hair_portion_list = list(
+				"bottom" = BOTTOM_DETAIL,
+				"middle" = MIDDLE_DETAIL,
+				"top" = TOP_DETAIL
+			)
+
+			var/hair_portion_selected = hair_portion_list[src.hair_portion]
+			var/datum/customization_style/new_hairstyle = null
+			var/all_hairs = null
+
+			if (istype(src, /datum/component/barber/shave))
+				all_hairs = concrete_typesof(/datum/customization_style/beard) + concrete_typesof(/datum/customization_style/moustache) + concrete_typesof(/datum/customization_style/sideburns) + concrete_typesof(/datum/customization_style/eyebrows)
+			else
+				all_hairs = concrete_typesof(/datum/customization_style/hair)
+
+			for (var/datum/customization_style/iteration as anything in all_hairs)
+				var/datum/customization_style/style = new iteration
+				if (style.id == params["style_id"])
+					new_hairstyle = style
+
+			if(istype(src, /datum/component/barber/shave))
+				actions.start_and_wait(new/datum/action/bar/barber/shave(src.barbee, src.barber, get_barbery_conditions(src.barbee, src.barber), new_hairstyle, hair_portion_selected), src.barber)
+			else
+				actions.start_and_wait(new/datum/action/bar/barber/haircut(src.barbee, src.barber, get_barbery_conditions(src.barbee, src.barber), new_hairstyle, hair_portion_selected), src.barber)
+
+			if (!barber)
+				return // If there's no barber, it's safe to say we've been disposed of
+
+			src.new_AH.CopyOther(src.barbee.bioHolder.mobAppearance)
+			src.preview.preview_thing.appearance = src.barbee.appearance
+			src.preview.update_appearance(src.new_AH)
+			return TRUE
+
+		if("update_preview")
+			switch(params["action"])
+				if("new_hair")
+					var/datum/customization_style/new_hairstyle = new /datum/customization_style/none // If we don't find any styles, we are probably trying to use the "none" style.
+					var/all_hairs = null
+					if (istype(src, /datum/component/barber/shave))
+						all_hairs = concrete_typesof(/datum/customization_style/beard) + concrete_typesof(/datum/customization_style/moustache) + concrete_typesof(/datum/customization_style/sideburns) + concrete_typesof(/datum/customization_style/eyebrows)
+					else
+						all_hairs = concrete_typesof(/datum/customization_style/hair)
```

Comment:
Same here, this should be handled with inheritance rather than a copy pasted snippet (and all the other instances of this being duplicated)